### PR TITLE
Generated reports

### PR DESCRIPTION
**Pull request by nadjajovanovic**

### CODE REVIEW

File: TimeSheet/src/main/java/projekat/controller/ReportController.java [hunk 1]
```diff
@@ -5,59 +5,32 @@
...
-	public ReportController(ReportRepository reportRepository) {
-		this.reportRepository = reportRepository;
+	public ReportController(ReportService reportService) {
+		this.reportService = reportService;
 	}
 	
 	@GetMapping("report")
-	public Collection<Report> getAllReports() {
-		return reportRepository.findAll();
+	public ResponseEntity<Collection<TimeSheetEntry>> getAllReports(@RequestBody Report report) {
```
Comment (igajic) [inline TimeSheet/src/main/java/projekat/controller/ReportController.java:27]: Please change return type to:

`List<TimeSheetEntry>`

**Comment by igajic** [comment]
Please add some test cases

File: TimeSheet/src/main/java/projekat/services/ReportService.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    public List<TimeSheetEntry> generateReport(Report report) {
+        final var allEntries = timeSheetEntryRepository.findAll();
+        final var filteredReports = allEntries.stream()
+                                                                    .filter(e -> report.getProjectid() == null || e.getProjectid() == report.getProjectid())
+                                                                    .filter(e -> report.getCategoryid() == null || e.getCategoryid() == report.getCategoryid())
+                                                                    .filter(e -> report.getClientid() == null || e.getClientid() == report.getClientid())
+                                                                    .filter(e -> report.getStartdate() == null || e.getEntryDate().after(report.getStartdate()))
+                                                                    .filter(e -> report.getEnddate() == null || e.getEntryDate().before(report.getEnddate()))
+                                                                    .collect(Collectors.toList());
```
Comment (igajic) [inline TimeSheet/src/main/java/projekat/services/ReportService.java:29]: With **Java 16** you can change

`.collect(Collectors.toList());`

into simple

`.toList();`

for older versions of Java, `collect(Collectors.toList())` is still way to go.

File: TimeSheet/src/main/java/projekat/models/Category.java [hunk 3]
```diff
@@ -30,8 +30,4 @@
 
 	private String categoryname;
 
-	//bi-directional many-to-one association to Report
```
[the remark this answers is not in the document]
Comment (nadjajovanovic) [reply TimeSheet/src/main/java/projekat/models/Category.java:33]: Sure, we will remove it

File: TimeSheet/src/main/java/projekat/services/ReportService.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    private TimeSheetEntryRepository timeSheetEntryRepository;
+
+    public ReportService(TimeSheetEntryRepository timeSheetEntryRepository) {
+        this.timeSheetEntryRepository = timeSheetEntryRepository;
+    }
+
+    public List<TimeSheetEntry> generateReport(Report report) {
+        final var allEntries = timeSheetEntryRepository.findAll();
+        final var filteredReports = allEntries.stream()
+                                                                    .filter(e -> report.getProjectid() == null || e.getProjectid() == report.getProjectid())
```
[the remark this answers is not in the document]
Comment (nadjajovanovic) [reply TimeSheet/src/main/java/projekat/services/ReportService.java:23]: All right. And for null, same .equals?

File: TimeSheet/src/test/java/projekat/util/BaseUT.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package projekat.util;
+
+import lombok.experimental.UtilityClass;
+import projekat.models.*;
+
+import java.util.Date;
+
+
+@UtilityClass
+public class BaseUT {
```
Comment (igajic) [inline TimeSheet/src/test/java/projekat/util/BaseUT.java:10]: Class **TimeSheetEntryControllerIntegrationTest** should extend **BaseUT**, methods within **BaseUT** should be protected

Please remove @UtilityClass

**Review by igajic — APPROVED** [review]

**Pull request merged**